[WIP] Use rolling update instead of delete/create when function exists

ericstoekl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This change is to perform in-place update of functions rather than delete/create when the function already exists.

If the function does not already exist and we get 404, create the function.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
#256 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Testing is improved in `proxy/deploy_test.go` by adding scaffolding with array-of-struct test cases.

We need to improve testing in this area by adding more test cases (hence why this is WIP).

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

BurtonR commented on proxy/deploy.go:
```diff
@@ -79,14 +95,14 @@ func DeployFunction(fprocess string, gateway string, functionName string, image
 	if err != nil {
 		fmt.Println(err)
-		return
+		return -1
```
Not sure if this matters, but the usage shows that you're returning a status code (in the function call, not the signature), so should this return 500 (`http.StatusInternalServerError`) instead?

Same comment for line 105 with the other `return -1`

Purely a nit-pick thing, just thinking about consistency for down the road

alexellis replied:
Yes what does -1 mean?

alexellis commented on proxy/deploy.go:
```diff
@@ -21,10 +21,26 @@ type FunctionResourceRequest struct {
+	constraints []string, update bool, secrets []string, labels map[string]string,
+	functionResourceRequest1 FunctionResourceRequest) {
+
+	statusCode := DeployFunctionImpl(fprocess, gateway, functionName, image, language, replace, envVars, network, constraints, update, secrets, labels, functionResourceRequest1)
```
Not keen on names like "Impl".. we can just call this `Deploy`? I saw we broke out a function for this - could we make the change one level higher? Does that even make sense?

ericstoekl replied:
By make the change one level higher -- do you mean having the retry logic (retry if 404 error occurs) in `commands/deploy.go`, in the `runDeploy()` function?

alexellis commented:
```
How Has This Been Tested?
```

@ericstoekl please tell us what end-to-end testing has been done on this.

@BurtonR have you played with it yet?

alexellis commented on proxy/deploy.go:
```diff
@@ -21,10 +21,25 @@ type FunctionResourceRequest struct {
+
+	statusCode := Deploy(fprocess, gateway, functionName, image, language, replace, envVars, network, constraints, update, secrets, labels, functionResourceRequest1)
+
+	if update == true && statusCode == http.StatusNotFound {
```
Just a *nit* here - can we avoid setting `update = false` and just pass `false` through the message signature? It will make debugging easier. Or you could create a new local variable as an alternative: `updateFunction := false`

Just to avoid destructively updating the value

alexellis commented on proxy/deploy.go:
```diff
@@ -21,10 +21,25 @@ type FunctionResourceRequest struct {
+	}
+}
+
+// Call FaaS server to deploy a new function
```
While here let's remove the comment here.

alexellis commented on proxy/deploy_test.go:
```diff
@@ -61,23 +60,31 @@ func Test_DeployFunction_Not2xx(t *testing.T) {
 	})
 
-	r := regexp.MustCompile(`(?m:Unexpected status: 404)`)
+	r := regexp.MustCompile(dpt.expectedOutput)
```
Can we change the name of `dpt` to something else?

BurtonR commented:
I've tested this by running the tests (all pass) as well as running through the commands as a normal user. 

`faas-cli deploy -f tester.yml` deploys a new function
`faas-cli deploy -f tester.yml` again, updates the function
`faas-cli deploy -f tester.yml --update=false --replace=true` replaces the function
All working as expected 👍 

_However_, the `faas-cli deploy --help` shows incorrect usage. This is not new for this PR, but something that's been lurking for a bit, I just forgot about it as I rarely use that any more.

In `commands/deploy.go`, lines 85-86, the help text for the deploy function is missing the second flag
Both flags are required as the CLI uses default values and must be explicitly overridden when passing the `--replace=true` flag
```
  faas-cli deploy -f ./samples.yml --replace=false
  faas-cli deploy -f ./samples.yml --update=true
```
should be:
```
  faas-cli deploy -f ./samples.yml --replace=false --update=true
  faas-cli deploy -f ./samples.yml --replace=true --update=false
```
or perhaps, remove the flags from the `update=true` as that's default behavior now.

alexellis commented:
@ericstoekl can you address the comment from @BurtonR please?

ericstoekl commented:
@alexellis just noticed there's been activity here, I'll have an update for this later today.

ericstoekl commented:
Just updated this, thanks to @BurtonR for helping with testing and suggestions. The latest commit addresses the comments.

@alexellis, I have tested this by running the tests as well as manually going through the scenario with a docker swarm deployment.

One thought I have is combining the `--update` and `--replace` flags into a single flag. What are your thoughts on this (Burton and/or Alex?)

BurtonR commented:
I agree, I think one flag would be more intuitive since we default one of them to true always and there's no way to check (that I'm aware of) if the flag was actually passed in.

So, just having the `--replace` flag would make a lot of sense. It updates automatically, but if the `--replace` flag is present, remove and deploy. Bonus points for also including a `-r` option :)

Also, I think that should go into a different PR since this one is focused on setting the default behavior.

alexellis commented on proxy/deploy_test.go:
```diff
@@ -14,43 +14,42 @@ import (
-	if !r.MatchString(stdout) {
-		t.Fatalf("Output not matched: %s", stdout)
-	}
+var DeployProxyTests = []DeployProxyTest{
```
The capital letter means this is exported outside the package. Was that the intention?

ericstoekl replied:
Was not the intention, I'll fix this.

alexellis commented on proxy/deploy_test.go:
```diff
@@ -61,23 +60,31 @@ func Test_DeployFunction_Not2xx(t *testing.T) {
 
+func Test_RunDeployProxyTests(t *testing.T) {
+	for _, tst := range DeployProxyTests {
+		t.Run(tst.title, func(t *testing.T) {
```
Just wanted to ask why you did it this way instead of our usual coding style of test-cases and an in-line struct?

ericstoekl replied:
I could change it so the test cases are defined in the `Test_RunDeployProxyTests()` method, but I like having `runDeployProxyTest()` as a separate function because it makes it easier to read

alexellis commented:
I think users will find the 404 message confusing. Please could you open a new PR so that we get this output?

When it exists:


```
Function already exists, attempting rolling-update.

Deployed.
URL: http://192.168.0.66:8080/function/cat

200 OK
```

When it doesn't:

```
Deployed.
URL: http://192.168.0.66:8080/function/cat

200 OK
```
